lib/cylc/remote.py

```diff
@@ -283,3 +296,544 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+                raise ValueError(
+                    "Threshold value '%s' of wrong type." % cutoff)
+            else:
+                valid_thresholds[measure] = cutoff
+        return valid_thresholds
+
+    @staticmethod
+    def selection_complete(host_list):
+        """Check for and address a list (of hosts) with length zero or one.
+
+           Check length of a list (of hosts); for zero items raise an error,
+           for one return that item, else (for multiple items) return False."""
+        if len(host_list) == 0:
+            raise EmptyHostList()
+        elif len(host_list) == 1:
+            return host_list[0]
+        else:
+            return False
+
+    def _trivial_choice(self, host_list, ignore_if_thresholds_prov=False):
+        """Address run host configuration that can be dealt with trivially.
+
+           If thresholds are not provided or set to be ignored, test if
+           'host_list' has length zero or one, else if the ranking method is
+           random; if so return an exception or host selected appropriately."""
+        if ignore_if_thresholds_prov and self.PARSED_THRESHOLDS:
+            return False
+        if self.selection_complete(host_list):
+            return self.selection_complete(host_list)
+        elif self.RANK_METHOD == 'random':
+            return choice(host_list)
+        else:
+            return False
+
+    def _process_get_host_metrics_cmd(self):
+        """Return 'get_host_metrics' command to run with only required options.
+
+           Return the command string to run 'cylc get host metric' with only
+           the required options given rank method and thresholds specified."""
+
+        # Convert config keys to associated command options. Ignore 'random'.
+        considerations = [self.RANK_METHOD] + self.PARSED_THRESHOLDS.keys()
+        opts = set()
+        for spec in considerations:
+            if spec.startswith("load"):
+                opts.add("--load")
+            elif spec.startswith("disk-space"):
+                opts.add("--disk-space=" + self.use_disk_path)
+            elif spec == "memory":
+                opts.add("--memory")
+
+        cmd = [self.CMD_BASE] + list(opts)
+        return " ".join(cmd)
+
+    def _remove_bad_hosts(self, cmd_with_opts, mock_stats=False):
+        """Return dictionary of 'good' hosts with their metric stats.
+
+           Run 'get-host-metrics' on each run host & store extracted stats for
+           'good' hosts only. Ignore 'bad' hosts whereby either metric data
+           cannot be accessed from the command or at least one metric value
+           does not pass a specified threshold."""
+        host_stats = {}
+        cmd = cmd_with_opts.split()
+        if mock_stats:  # Create fake data for unittest purposes (only).
+            host_stats = dict(mock_stats)  # Prevent mutable object issues.
+        else:
+            for host in self.HOSTS:
+                if host == 'localhost':
```

Comment:
I think this could potentially cause problems if one were to specify their localhost by name (rather than as `localhost`).

Here is a diff from #988:

```diff
diff --git a/lib/cylc/hostuserutil.py b/lib/cylc/hostuserutil.py
index b6ec6b27e..966f335ce 100644
--- a/lib/cylc/hostuserutil.py
+++ b/lib/cylc/hostuserutil.py
@@ -148,6 +148,8 @@ class HostUtil(object):
 
     def get_fqdn_by_host(self, target):
         """Return the fully qualified domain name of the target host."""
+        if target == 'localhost':
+            return socket.getfqdn()
         return self._get_host_info(target)[0]
 
     def get_user(self):
```
Which you can use to resolve the value of `localhost` like so:
```python
if get_fqdn_by_host(host) == get_fqdn_by_host('localhost'):
```

Reply:
Very good point & suggestion. Will amend shortly.

Reply:
It seems 'localhost' gets resolved through the *existing* ``HostUtil`` methods, without need for the included diff (see latest commit, & note the new unit sub-test checking ``get_fqdn_by_host('localhost')`` fails when checking for the unresolved equivalent ``.get('localhost', False)``.